quickwit/quickwit-common/src/rate_limited_tracing.rs

```diff
@@ -17,44 +17,121 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-#[macro_export]
-macro_rules! rate_limited_tracing {
-    ($log_fn:ident, limit_per_min=$limit:literal, $($args:tt)*) => {{
-        use ::std::sync::atomic::{AtomicU32, Ordering};
-        use ::std::sync::Mutex;
-        use ::std::time::{Instant, Duration};
-
-        static COUNT: AtomicU32 = AtomicU32::new(0);
-        // we can't build an Instant from const context, so we pinitialize with a None
-        static LAST_RESET: Mutex<Option<Instant>> = Mutex::new(None);
-
-        let count = COUNT.fetch_add(1, Ordering::Relaxed);
-        if count == 0 {
-            // this can only be reached the very 1st time we log
-            *LAST_RESET.lock().unwrap() = Some(Instant::now());
+// TODO coasetime has a recent() instead of now() which is essentially free (atomic read instead of
+// vdso call), but needs us to spawn a future/thread updating that value regularly
+
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use coarsetime::{Duration, Instant};
+
+#[derive(Clone, Copy)]
+struct Count {
+    generation: u32,
+    call_count: u32,
+}
+
+impl From<u64> for Count {
+    fn from(val: u64) -> Count {
+        Count {
+            generation: (val >> 32) as u32,
+            call_count: (val & ((1 << 32) - 1)) as u32,
         }
+    }
+}
+
+impl From<Count> for u64 {
+    fn from(count: Count) -> u64 {
+        ((count.generation as u64) << 32) + count.call_count as u64
+    }
+}
+
+/// Helper function used in [`rate_limited_tracing`] to determine if this line should log,
+/// and update the related counters.
+pub fn should_log<F: Fn() -> Instant>(
+    count_atomic: &AtomicU64,
+    last_reset_atomic: &AtomicU64,
+    limit: u32,
+    now: F,
+) -> bool {
+    //  count_atomic is treated as 2 u32: upper bits count "generation", lower bits count number of
+    //  calls since LAST_RESET. We assume there won't be 2**32 calls to this log in ~60s.
+    //  Generation is free to wrap arround.
+
+    let count = count_atomic.fetch_add(1, Ordering::Acquire);
```

Comment:
no shadowing please.
```suggestion
    // Because the lower 32 bits are storing the log count, we can
    // increment the entire u64 to record this log call.
    let logsite_data_as_u64: u64 = count_atomic.fetch_add(1, Ordering::Acquire);
```

Reply:
`let count: Count = count.into();`

this is shadowing.